packages/nocodb/src/lib/version-upgrader/ncFilterUpgrader_0104004.ts

```diff
@@ -0,0 +1,380 @@
+        filter.id,
+        {
+          ...filter,
+          comparison_op: 'anyof',
+          value: newFilterValue,
+        },
+        ncMeta
+      )
+    );
+  } else if (filter.comparison_op === 'neq') {
+    // migrate to `doesn't contain all of`
+    actions.push(
+      await Filter.update(
+        filter.id,
+        {
+          ...filter,
+          comparison_op: 'nanyof',
+          value: newFilterValue,
+        },
+        ncMeta
+      )
+    );
+  }
+  return actions;
+};
+
+const migrateToCheckboxFilter = async (filter, actions: any[], ncMeta) => {
+  if (['empty', 'null'].includes(filter.comparison_op)) {
+    // migrate to not checked
+    actions.push(
+      await Filter.update(
+        filter.id,
+        {
+          ...filter,
+          comparison_op: 'notchecked',
+        },
+        ncMeta
+      )
+    );
+  } else if (['notempty', 'notnull'].includes(filter.comparison_op)) {
+    //  migrate to checked
+    actions.push(
+      await Filter.update(
+        filter.id,
+        {
+          ...filter,
+          comparison_op: 'checked',
+        },
+        ncMeta
+      )
+    );
+  } else if (filter.comparison_op === 'eq') {
+    if (['true', 'True', '1', 'T', 'Y'].includes(filter.value)) {
+      // migrate to checked
+      actions.push(
+        await Filter.update(
+          filter.id,
+          {
+            ...filter,
+            comparison_op: 'checked',
+            value: '',
+          },
+          ncMeta
+        )
+      );
+    } else if (['false', 'False', '0', 'F', 'N'].includes(filter.value)) {
+      // migrate to notchecked
+      actions.push(
+        await Filter.update(
+          filter.id,
+          {
+            ...filter,
+            comparison_op: 'notchecked',
+            value: '',
+          },
+          ncMeta
+        )
+      );
+    } else {
+      // invalid value - good to delete
+      actions.push(await Filter.delete(filter.id, ncMeta));
+    }
+  } else if (filter.comparison_op === 'neq') {
+    if (['false', 'False', '0', 'F', 'N'].includes(filter.value)) {
+      // migrate to checked
+      actions.push(
+        await Filter.update(
+          filter.id,
+          {
+            ...filter,
+            comparison_op: 'checked',
+            value: '',
+          },
+          ncMeta
+        )
+      );
+    } else if (['true', 'True', '1', 'T', 'Y'].includes(filter.value)) {
+      // migrate to not checked
+      actions.push(
+        await Filter.update(
+          filter.id,
+          {
+            ...filter,
+            comparison_op: 'notchecked',
+            value: '',
+          },
+          ncMeta
+        )
+      );
+    } else {
+      // invalid value - good to delete
+      actions.push(await Filter.delete(filter.id, ncMeta));
+    }
+  }
+  return actions;
+};
+
+async function migrateFilters(ncMeta: NcMetaIO) {
+  const filters = await ncMeta.metaList2(null, null, MetaTable.FILTER_EXP);
+  let actions = [];
```

Comment:
In this function, we are deleting and changing the behavior of existing filters, right? So it going to be a breaking change and the result may vary?

Reply:
Yes. Since some filter types are not appropriate to some column types, they will be deleted. For some filter types, we will migrate to others instead. The breaking change summary is listed in the above PR summary. 

For the new filter matrix, it has been recorded [here](https://docs.google.com/spreadsheets/d/e/2PACX-1vTpCNKtA-szaXUKJEO5uuSIRnzUOK793MKnyBz9m2rQcwn7HqK19jPHeER-IIRWH9X56J78wfxXZuuv/pubhtml?gid=427284630&amp;single=true&amp;widget=true&amp;headers=false).